lib/lattice_observer/observed/event_processor.ex

```diff
@@ -193,33 +195,63 @@ defmodule LatticeObserver.Observed.EventProcessor do
   def record_heartbeat(l = %Lattice{}, source_host, stamp, data) do
     labels = Map.get(data, "labels", %{})
     friendly_name = Map.get(data, "friendly_name", "")
+    annotations = Map.get(data, "annotations", %{})
+    spec = Map.get(annotations, @annotation_app_spec, "")
+
+    # Heartbeats are now considered authoritative, so the previously stored
+    # actor and provider list are wiped prior to recording the heartbeat, but the
+    # host (and its "first seen" time) are not.
+    l =
+      %Lattice{
+        l
+        | actors:
+            l.actors
+            |> Enum.map(fn {k, v} ->
+              {k,
+               %Actor{
+                 v
+                 | instances: v.instances |> Enum.reject(fn i -> i.host_id == source_host end)
```

Comment:
So, we reject all of the instances of these actors and providers and then add them all back, is that correct?

I have some concern about this getting expensive when monitoring many lattices, though I doubt it would really degrade performance meaningfully. The only thing I could think of is comparing instance IDs but that feels a little scary

Reply:
This filter is pretty fast. This basically just means we dump the actors and providers from a host when we get its heartbeat. This is pretty much the definition of making the heartbeat authoritative.

Reply:
I've been down the deep, dark road of doing "diff comparisons" in order to try and optimize the pruning, and most of the time it's way more error prone than the performance bump justifies.

Reply:
That makes sense, and even if we have to change in the future this will be easy to nail down 😄